Add travis build

erikd opened the pull request:
Only builds with GHC 7.10 and 8.0 for now. Can't build in Travis with 7.8.4 because the compiler available hasn't been built with profiling and hence the tests fail. Building with 8.2 is also busted and will be fixed in an upcoming PR.

@haskell-mafia

erikd committed:
Update submodules so it can build in travis

All the submodules (and their submodules) need to be accessible via
https (ie no ssh access).

erikd committed:
Add .travis.yml

erikd committed:
Mafia.Script: Update SUBMODULE handling

* Allow orgs and repos with dashes (eg haskell-mafia org).
* Allow an optional submodule location to be specified eg:

   {-# SUBMODULE https://gitlab.com user/repo@commit #-}

If the location ("https://gitlab.com" in the above) is not present
it is assumed to be "git@github.com:" to maintain backwards
compatibility.

Also add a round-trip test for the SUBMODULE pragma parser.

erikd committed:
Mafia.Hoogle: Use installed hoogle if found

This saves building hoogle from source every time this test is run.

nhibberd commented on src/Mafia/Script.hs:
```diff
@@ -136,7 +136,7 @@ parseScript path source =
 renderSubmoduleUrl :: Submodule -> Text
 renderSubmoduleUrl s =
-  "git@github.com:" <> submoduleUser s <> "/" <> submoduleRepo s
+  "https://github.com/" <> submoduleUser s <> "/" <> submoduleRepo s
```
Isn't this a breaking change for mafia scripts? can the protocol be added to the Pragma in a compatible way? 

e.g.

```
data Submodule =
  Submodule {
      submoduleProtocol :: !Protocol
    , submoduleUser :: !Text
    , submoduleRepo :: !Text
    , submoduleCommit :: !(Maybe Text)
    } deriving (Eq, Ord, Show)

Current implicit behaviour:
{-# SUBMODULE ambiata/p@xxx #-} -> Submodule Ssh "ambiata" "p" "xxx"

Explicit protocol:
{-# SUBMODULE ssh ambiata/p@xxx #-} -> Submodule Ssh "ambiata" "p" "xxx"

{-# SUBMODULE https ambiata/p@xxx #-} -> Submodule Https "ambiata" "p" "xxx"
```

I am sure there is a better way of describing the protocols, just first thing that came to mind that would keep compatibility

erikd replied:
Actually, why not make the protocol and the server explicit like:
```
{-# SUBMODULE git@github.com ambiata/p@xxx #-}
{-# SUBMODULE https://gitlab.com ambiata/x@xxx #-}
```
so that other services like Gitlab and Bitbucket can be used.

Its harder to make this backwards compatible though.

charleso replied:
> Actually, why not make the protocol and the server explicit like:

I'm actually 100% on board with this, I never liked hard coding to github. That said, my suggestion is that you should also fall back to the previous behaviour to remain backwards compatible (as well as keeping with Jacob's original intention)

erikd committed:
Fix cli tests to run in travis

Drop dependencies on github repos not belonging to the haskell-mafia
organisation. The haskell-mafia org now has two test repos specifically
designed to test mafia but have minimal dependencies so they can be
built quickly.

charleso commented on src/Mafia/Script.hs:
```diff
@@ -5,11 +5,20 @@ module Mafia.Script (
   , runScript
+
+  -- For testing.
+  , Pragma (..)
```
This whole script change seems unrelated to building in travis. Should it be it's own PR?

erikd committed:
script/mafia: Switch to haskell-mafia

tmcgilchrist reviewed (approved):
Getting this building on travis is a good first step. 
We'll inline / remove the submodules in later PRs
